blocks/algorithms/__init__.py

```diff
@@ -180,7 +180,11 @@ class GradientDescent(DifferentiableCostMinimizer):
 
     """
     def __init__(self, step_rule=None, gradients=None, **kwargs):
-        super(GradientDescent, self).__init__(**kwargs)
+        if gradients and 'params' not in kwargs:
```

Comment:
This indeed looks confusing -- maybe it should read:

`if gradients is not Null and 'params' not in kwargs:`

I was also wondering: Maybe DifferentiableCostMinimizer shouldn't populate `params` at all? Identifying parameter variables that can be optimized is maybe something a more specialized subclass should do? E.g. GradientDescent. That would simplify/streamline the code here a bit.